sensor/common/clusterentities/store_endpoints_test.go

```diff
@@ -0,0 +1,289 @@
+				// after tick 2: endpoint should be marked as historical
+				2: {expectDeployment80("10.0.0.1", "depl1", "http", history)},
+				// after tick 3: endpoint should be no longer considered historical
+				3: {expectDeployment80("10.0.0.1", "depl1", "http", theMap)},
+			},
+		},
+		"Re-add deployment previously marked as historical after history expired": {
+			numTicksToRemember: 2,
+			entityUpdates: map[int][]eUpdate{
+				0: {
+					{
+						deploymentID: "depl1",
+						containerID:  "pod1",
+						ipAddr:       "10.0.0.1",
+						port:         80,
+						portName:     "http",
+						incremental:  true,
+					},
+				},
+				4: {
+					{
+						deploymentID: "depl1",
+						containerID:  "pod1",
+						ipAddr:       "10.0.0.1",
+						port:         80,
+						portName:     "http",
+						incremental:  true,
+					},
+				},
+			},
+			operationAfterTick: map[int]operation{1: deleteDeployment1},
+			lookupResultsAfterTick: map[int][]expectation{
+				// before tick 1: endpoint should be added immediately
+				0: {expectDeployment80("10.0.0.1", "depl1", "http", theMap)},
+				// after tick 1: no change expected
+				1: {expectDeployment80("10.0.0.1", "depl1", "http", theMap)},
+				// deletion happens here
+				// after tick 2: endpoint should be marked as historical
+				2: {expectDeployment80("10.0.0.1", "depl1", "http", history)},
+				// after tick 3: endpoint should still be in the history
+				3: {expectDeployment80("10.0.0.1", "depl1", "http", history)},
+				// after tick 4: endpoint should be no longer considered historical
+				4: {expectDeployment80("10.0.0.1", "depl1", "http", theMap)},
+			},
+		},
+		"IP changing the owner deployment should populate history with previous IP": {
+			numTicksToRemember: 1,
+			entityUpdates: map[int][]eUpdate{
+				0: {
+					{
+						deploymentID: "depl1",
+						containerID:  "pod1",
+						ipAddr:       "10.0.0.1",
+						port:         80,
+						portName:     "http",
+						incremental:  true,
+					},
+				},
+				2: {
+					{
+						deploymentID: "depl2",
+						containerID:  "pod2",
+						ipAddr:       "10.0.0.1",
+						port:         80,
+						portName:     "http",
+						incremental:  false, // overwrite
+					},
+				},
+			},
+			operationAfterTick: map[int]operation{},
+			lookupResultsAfterTick: map[int][]expectation{
+				// before tick 1: endpoint should be added immediately
+				0: {expectDeployment80("10.0.0.1", "depl1", "http", theMap)},
+				// after tick 1: no change expected
+				1: {expectDeployment80("10.0.0.1", "depl1", "http", theMap)},
+				// after tick 1: endpoint should be overwritten and old IP placed in history
+				2: {
+					expectDeployment80("10.0.0.1", "depl1", "http", history),
+					expectDeployment80("10.0.0.1", "depl2", "http", theMap),
+				},
+				// after tick 2: history expires
+				3: {
+					expectDeployment80("10.0.0.1", "depl1", "http", nowhere),
+					expectDeployment80("10.0.0.1", "depl2", "http", theMap),
+				},
+			},
+		},
+	}
```

Comment:
If a deployment updates multiple times within a tick, what's the expected behavior? I've seen that current tick rate setting is 30s and I don't know how realistic this scenario is, but codifying the behavior and testing it might be a good idea.

Reply:
Yes, this is a great idea, and we partially have it in the tests - check a test case named `Memory enabled with reset after tick 1 should make historical IPs be remembered` - it does two updates in one tick.

The default tick is 30s and we may have really a lot of updates between the ticks. Notice that if history is enabled (default setting), then we never delete anything from the stores and we always wait for a tick, because only on tick we can remove an entry completely. Between the ticks, we may move items between `theMap` and history, but the information is not dropped.